File: diffusionx/src/simulation/levy.rs
//! Levy process simulation
//! For Brownian motion, see [`crate::simulation::bm`].

use crate::{
    SimulationError, XResult,
    random::stable,
    simulation::{Pair, Simulation, Stochastic, functional::FirstPassageTime},
    utils::cumsum,
};
use rayon::prelude::*;

/// Levy process simulation
///
/// This struct represents a Levy process simulation.
///
/// # Fields
///
/// * `start_position` - The starting position of the Levy process.
/// * `alpha` - The stability index of the Levy process.
#[derive(Debug, Clone)]
pub struct Levy {
    start_position: f64,
    alpha: f64,
}

impl Levy {
    /// Create a new Levy process simulation
    ///
    /// # Arguments
    ///
    /// * `start_position` - The starting position of the Levy process.
    /// * `alpha` - The stability index of the Levy process.
    pub fn new(start_position: impl Into<f64>, alpha: impl Into<f64>) -> XResult<Self> {
        let start_position = start_position.into();
        let alpha = alpha.into();
        if alpha <= 0.0 || alpha > 2.0 {
            return Err(SimulationError::InvalidParameters(
                "alpha must be in the range (0, 2]".to_string(),
            )
            .into());
        }
        Ok(Self {
            start_position,
            alpha,
        })
    }

    /// Get the starting position of the Levy process simulation
    pub fn start_position(&self) -> f64 {
        self.start_position
    }

    /// Get the stability index of the Levy process simulation
    pub fn index(&self) -> f64 {
        self.alpha
    }

    /// Get the first passage time of the Levy process simulation
    /// 
    /// # Arguments
    /// 
    /// * `domain` - The domain of the Levy process simulation.
    /// * `max_duration` - The maximum duration of the Levy process simulation.
    /// * `time_step` - The time step of the Levy process simulation.
    ///
    /// # Returns
    ///
    /// `Option<f64>`
    /// * None if the first passage time is not found within the maximum duration.
    /// * A f64 representing the first passage time of the simulation.
    ///
    /// # Example
    ///
    /// ```rust
    /// let levy = Levy::new(0.0, 1.5).unwrap();
    /// let params = 0.1;
    /// let (t, x) = levy.simulate(params).unwrap();
    /// ```
    pub fn fpt(
        &self,
        domain: (impl Into<f64>, impl Into<f64>),
        max_duration: impl Into<f64>,
        time_step: f64,
    ) -> XResult<Option<f64>> {
        let fpt = FirstPassageTime::new(self, domain)?;
        fpt.simulate(max_duration, time_step)
    }
}

impl Stochastic for Levy {}

/// impl `Simulation` trait for Levy process
impl Simulation for Levy {
    /// Simulate Levy process
    ///
    /// # Arguments
    ///
    /// * `time_step` - The time step of the Levy process simulation.
    ///
    /// # Returns
    ///
    /// A tuple containing the time and the position of the Levy process simulation.
    ///
    /// # Example
    ///
    /// ```rust
    /// let levy = Levy::new(0.0, 1.5).unwrap();
    /// let params = 0.1;
    /// let (t, x) = levy.simulate(params).unwrap();
    /// ```
    fn simulate(&self, duration: impl Into<f64>, time_step: f64) -> XResult<Pair> {
        if time_step <= 0.0 {
            return Err(SimulationError::InvalidParameters(
                "time_step must be positive".to_string(),
            )
            .into());
        }
        let duration = duration.into();
        if duration <= 0.0 {
            return Err(SimulationError::InvalidParameters(
                "duration must be positive".to_string(),
            )
            .into());
        }
        simulate_levy(self.start_position, self.alpha, duration, time_step)
    }
}

/// Simulate Levy process
///
/// This function simulates Levy process.
///
/// # Arguments
///
/// * `start_position` - The starting position of the Levy process.  
/// * `alpha` - The stability index of the Levy process.
/// * `tau` - The time step of the Levy process.
/// * `duration` - The duration of the Levy process.
///
/// # Returns
///
/// The result of the Levy process simulation.   
///
/// # Example
///
/// ```rust
/// let levy = Levy::new(0.0, 1.5).unwrap();
/// let params = 0.1;
/// let (t, x) = levy.simulate(params).unwrap();
/// ```
pub fn simulate_levy(
    start_position: impl Into<f64>,
    alpha: impl Into<f64>,
    duration: impl Into<f64>,
    time_step: f64,
) -> XResult<(Vec<f64>, Vec<f64>)> {
    let start_position = start_position.into();
    let alpha = alpha.into();
    let duration = duration.into();
    let num_steps = (duration / time_step).ceil() as usize;
    let t = (0..=num_steps)
        .into_par_iter()
        .map(|i| time_step * i as f64)
        .collect::<Vec<_>>();
    let noise = stable::standard_rands(alpha, 0.0, num_steps)?
        .into_par_iter()
        .map(|x| x * time_step.powf(1.0 / alpha))
        .collect::<Vec<_>>();
    let x = cumsum(start_position, &noise);
    Ok((t, x))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simulate_levy() {
        let levy = Levy::new(10.0, 1.5).unwrap();
        let time_step = 0.1;
        let duration = 1.0;
        let (t, x) = levy.simulate(duration, time_step).unwrap();
        println!("t: {:?}", t);
        println!("x: {:?}", x);
    }

    #[test]
    fn test_fpt() {
        let levy = Levy::new(0.0, 1.5).unwrap();
        let time_step = 0.1;
        let fpt = levy.fpt((-1.0, 1.0), 1000.0, time_step).unwrap();
        println!("fpt: {:?}", fpt);
    }

    #[test]
    fn test_send_sync() {
        fn assert_send_sync<T: Send + Sync>() {}
        assert_send_sync::<Levy>();
    }
}

File: py-diffusionx/python/diffusionx/simulation/levy.py
from diffusionx import _core
from typing import Union
from .basic import StochasticProcess, Trajectory
from .utils import check_transform
import numpy as np


real = Union[float, int]


class Levy(StochasticProcess):
    def __init__(
        self,
        alpha: real,
        start_position: real = 0.0,
    ):
        """
        Initialize a Lévy process object.

        Args:
            alpha (real): The alpha parameter of the Lévy process, the value must be in the range (0, 2].
            start_position (real, optional): Starting position of the Lévy process. Defaults to 0.0.

        Raises:
            ValueError: If duration is not positive.
            ValueError: If diffusion coefficient is not positive.
            ValueError: If value is not a number.

        Returns:
            Bm: A Brownian motion object.
        """
        start_position = check_transform(start_position)
        alpha = check_transform(alpha)
        if alpha <= 0 or alpha > 2:
            raise ValueError("alpha must be in the range (0, 2]")

        self.start_position = start_position
        self.alpha = alpha

    def __call__(self, duration: real) -> Trajectory:
        return Trajectory(self, duration)

    def simulate(
        self, duration: real, step_size: real = 0.01
    ) -> tuple[np.ndarray, np.ndarray]:
        """
        Simulate the Lévy process.

        Args:
            duration (real): The duration of the Lévy process.
            step_size (real, optional): Step size of the Lévy process. Defaults to 0.01.

        Returns:
            tuple[np.ndarray, np.ndarray]: A tuple containing the times and positions of the Lévy process.
        """
        step_size = check_transform(step_size)
        if step_size <= 0:
            raise ValueError("step_size must be positive")
        return _core.levy_simulate(
            self.start_position,
            self.alpha,
            duration,
            step_size,
        )

    def fpt(
        self,
        domain: tuple[real, real],
        step_size: real = 0.01,
        max_duration: real = 1000,
    ):
        """
        Calculate the first passage time of the Brownian motion.

        Args:
            domain (tuple[real, real]): The domain of the Brownian motion.
            step_size (real, optional): Step size of the Brownian motion. Defaults to 0.01.

        Returns:
            real: The first passage time of the Brownian motion.
        """
        step_size = check_transform(step_size)
        a = check_transform(domain[0])
        b = check_transform(domain[1])
        max_duration = check_transform(max_duration)
        return _core.levy_fpt(
            self.start_position,
            self.alpha,
            step_size,
            (a, b),
            max_duration,
        )

    def occupation_time(
        self,
        domain: tuple[real, real],
        duration: real,
        step_size: real = 0.01,
    ):
        """
        Calculate the occupation time of the Lévy process.

        Args:
            domain (tuple[real, real]): The domain of the Lévy process.
            duration (real): The duration of the Lévy process.
            step_size (real, optional): Step size of the Lévy process. Defaults to 0.01.

        Returns:
            real: The occupation time of the Lévy process.
        """
        step_size = check_transform(step_size)
        duration = check_transform(duration)
        if duration <= 0:
            raise ValueError("duration must be positive")
        a = check_transform(domain[0])
        b = check_transform(domain[1])
        if a >= b:
            raise ValueError("domain must be a valid interval")
        return _core.levy_occupation_time(
            self.start_position,
            self.alpha,
            step_size,
            (a, b),
            duration,
        )
